In [2]:
# Update sklearn to prevent version mismatches
#!pip install sklearn --upgrade

In [3]:
# install joblib. This will be used to save your model. 
# Restart your kernel after installing 
#!pip install joblib

In [2]:
import pandas as pd
import numpy as np

Import our white and red wine data from the csvs

# Read the CSV and Perform Basic Data Cleaning

In [4]:
red_wine_df = pd.read_csv("Resources/winequality-red.csv", sep=';')
white_wine_df = pd.read_csv("Resources/winequality-white.csv", sep=';')
white_wine_df.info()
red_wine_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4898 entries, 0 to 4897
Data columns (total 12 columns):
fixed acidity           4898 non-null float64
volatile acidity        4898 non-null float64
citric acid             4898 non-null float64
residual sugar          4898 non-null float64
chlorides               4898 non-null float64
free sulfur dioxide     4898 non-null float64
total sulfur dioxide    4898 non-null float64
density                 4898 non-null float64
pH                      4898 non-null float64
sulphates               4898 non-null float64
alcohol                 4898 non-null float64
quality                 4898 non-null int64
dtypes: float64(11), int64(1)
memory usage: 459.3 KB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1599 entries, 0 to 1598
Data columns (total 12 columns):
fixed acidity           1599 non-null float64
volatile acidity        1599 non-null float64
citric acid             1599 non-null float64
residual sugar          1599 non-null float64

## Unbalanced (Denormalized) Data for White Wine Data (4898 data points) and Red Wine Data (1599 data points) 

We want to use all the white and red wine data baseline condition we can use:

In [5]:
#if we want our only features of the model to consist of all white and red wine data, we use this data frame:
baseline_red_df = red_wine_df.copy()
baseline_white_df = white_wine_df.copy()
baseline_wine_df = pd.concat([baseline_red_df, baseline_white_df])

In [23]:
baseline_wine_df.groupby('quality').mean()

fixed acidity  volatile acidity  citric acid  residual sugar  \
quality                                                                 
3             7.853333          0.517000     0.281000        5.140000   
4             7.288889          0.457963     0.272315        4.153704   
5             7.326801          0.389614     0.307722        5.804116   
6             7.177257          0.313863     0.323583        5.549753   
7             7.128962          0.288800     0.334764        4.731696   
8             6.835233          0.291010     0.332539        5.382902   
9             7.420000          0.298000     0.386000        4.120000   

         chlorides  free sulfur dioxide  total sulfur dioxide   density  \
quality                                                                   
3         0.077033            39.216667            122.033333  0.995744   
4         0.060056            20.636574            103.432870  0.994833   
5         0.064666            30.237371            120.839102  0.995849   
6         0.054157            31.165021            115.410790  0.994558   
7         0.045272            30.422150            108.498610  0.993126   
8         0.041124            34.533679            117.518135  0.992514   
9         0.027400            33.400000            116.000000  0.991460   

               pH  sulphates    alcohol  
quality                                  
3        3.257667   0.506333  10.215000  
4        3.231620   0.505648  10.180093  
5        3.212189   0.526403   9.837783  
6        3.217726   0.532549  10.587553  
7        3.228072   0.547025  11.386006  
8        3.223212   0.512487  11.678756  
9        3.308000   0.466000  12.180000

In [9]:
#Suppose we want to take specific ranges in only_balanced_df, which does not include colors
# range_colors_df = dummy_colors_df.loc[\
# (dummy_colors_df["quality"] <= 8 ) & \
# (dummy_colors_df["quality"] >= 3 )  \
# ] 


In [8]:
bin_balanced_df = baseline_wine_df.copy()
bins = [0, 4, 5, 6, 10]
group_names = [0, 1, 2, 3]
bin_balanced_df["bin_quality"] = pd.cut(bin_balanced_df["quality"], bins, labels=group_names)
bin_balanced_df.head(30)

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0             7.4             0.700         0.00             1.9      0.076   
1             7.8             0.880         0.00             2.6      0.098   
2             7.8             0.760         0.04             2.3      0.092   
3            11.2             0.280         0.56             1.9      0.075   
4             7.4             0.700         0.00             1.9      0.076   
5             7.4             0.660         0.00             1.8      0.075   
6             7.9             0.600         0.06             1.6      0.069   
7             7.3             0.650         0.00             1.2      0.065   
8             7.8             0.580         0.02             2.0      0.073   
9             7.5             0.500         0.36             6.1      0.071   
10            6.7             0.580         0.08             1.8      0.097   
11            7.5             0.500         0.36             6.1      0.071   
12            5.6             0.615         0.00             1.6      0.089   
13            7.8             0.610         0.29             1.6      0.114   
14            8.9             0.620         0.18             3.8      0.176   
15            8.9             0.620         0.19             3.9      0.170   
16            8.5             0.280         0.56             1.8      0.092   
17            8.1             0.560         0.28             1.7      0.368   
18            7.4             0.590         0.08             4.4      0.086   
19            7.9             0.320         0.51             1.8      0.341   
20            8.9             0.220         0.48             1.8      0.077   
21            7.6             0.390         0.31             2.3      0.082   
22            7.9             0.430         0.21             1.6      0.106   
23            8.5             0.490         0.11             2.3      0.084   
24            6.9             0.400         0.14             2.4      0.085   
25            6.3             0.390         0.16             1.4      0.080   
26            7.6             0.410         0.24             1.8      0.080   
27            7.9             0.430         0.21             1.6      0.106   
28            7.1             0.710         0.00             1.9      0.080   
29            7.8             0.645         0.00             2.0      0.082   

    free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                  11.0                  34.0   0.9978  3.51       0.56   
1                  25.0                  67.0   0.9968  3.20       0.68   
2                  15.0                  54.0   0.9970  3.26       0.65   
3                  17.0                  60.0   0.9980  3.16       0.58   
4                  11.0                  34.0   0.9978  3.51       0.56   
5                  13.0                  40.0   0.9978  3.51       0.56   
6                  15.0                  59.0   0.9964  3.30       0.46   
7                  15.0                  21.0   0.9946  3.39       0.47   
8                   9.0                  18.0   0.9968  3.36       0.57   
9                  17.0                 102.0   0.9978  3.35       0.80   
10                 15.0                  65.0   0.9959  3.28       0.54   
11                 17.0                 102.0   0.9978  3.35       0.80   
12                 16.0                  59.0   0.9943  3.58       0.52   
13                  9.0                  29.0   0.9974  3.26       1.56   
14                 52.0                 145.0   0.9986  3.16       0.88   
15                 51.0                 148.0   0.9986  3.17       0.93   
16                 35.0                 103.0   0.9969  3.30       0.75   
17                 16.0                  56.0   0.9968  3.11       1.28   
18                  6.0                  29.0   0.9974  3.38       0.50   
19                 17.0                  56.0   0.9969

In [10]:
bin_balanced_df.groupby("bin_quality").count()

fixed acidity  volatile acidity  citric acid  residual sugar  \
bin_quality                                                                 
0                      246               246          246             246   
1                     2138              2138         2138            2138   
2                     2836              2836         2836            2836   
3                     1277              1277         1277            1277   

             chlorides  free sulfur dioxide  total sulfur dioxide  density  \
bin_quality                                                                  
0                  246                  246                   246      246   
1                 2138                 2138                  2138     2138   
2                 2836                 2836                  2836     2836   
3                 1277                 1277                  1277     1277   

               pH  sulphates  alcohol  quality  
bin_quality                                     
0             246        246      246      246  
1            2138       2138     2138     2138  
2            2836       2836     2836     2836  
3            1277       1277     1277     1277

In [13]:
bad_df = bin_balanced_df.loc[bin_balanced_df['bin_quality']==0,:]
good_df = bin_balanced_df.loc[bin_balanced_df['bin_quality']==1,:]
great_df = bin_balanced_df.loc[bin_balanced_df['bin_quality']==2,:]
terrific_df = bin_balanced_df.loc[bin_balanced_df['bin_quality']==3,:]

In [14]:
normalized_good_df = good_df.sample(n=246,random_state=42)

In [15]:
normalized_great_df = great_df.sample(n=246,random_state=42)

In [16]:
normalized_terrific_df = terrific_df.sample(n=246,random_state=42)

In [17]:
normal_df = pd.concat([bad_df,normalized_good_df,normalized_great_df,normalized_terrific_df])
normdf = normal_df.drop(columns = 'quality')
normdf

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
18              7.4             0.590         0.08            4.40      0.086   
38              5.7             1.130         0.09            1.50      0.172   
41              8.8             0.610         0.30            2.80      0.088   
45              4.6             0.520         0.15            2.10      0.054   
73              8.3             0.675         0.26            2.10      0.084   
...             ...               ...          ...             ...        ...   
689             6.7             0.170         0.37            2.00      0.039   
456             6.3             0.230         0.31            1.50      0.022   
1830            7.7             0.440         0.24           11.20      0.031   
2776            7.0             0.360         0.32           10.50      0.045   
1127            6.4             0.125         0.29            5.85      0.042   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
18                    6.0                  29.0   0.9974  3.38       0.50   
38                    7.0                  19.0   0.9940  3.50       0.48   
41                   17.0                  46.0   0.9976  3.26       0.51   
45                    8.0                  65.0   0.9934  3.90       0.56   
73                   11.0                  43.0   0.9976  3.31       0.53   
...                   ...                   ...      ...   ...        ...   
689                  34.0                 125.0   0.9922  3.26       0.60   
456                  11.0                  82.0   0.9892  3.30       0.40   
1830                 41.0                 167.0   0.9948  3.12       0.43   
2776                 35.0                 135.0   0.9935  3.09       0.33   
1127                 24.0                  99.0   0.9920  3.23       0.32   

      alcohol bin_quality  
18        9.0           0  
38        9.8           0  
41        9.3           0  
45       13.1           0  
73        9.2           0  
...       ...         ...  
689      10.8           3  
456      12.9           3  
1830     11.3           3  
2776     11.6           3  
1127     12.0           3  

[984 rows x 12 columns]

# Select your features (columns)

In [18]:
# Set features. Drop quality, y value. 
X = normdf.drop(columns=['bin_quality'])
X.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
18            7.4             0.590         0.08             4.4      0.086   
38            5.7             1.130         0.09             1.5      0.172   
41            8.8             0.610         0.30             2.8      0.088   
45            4.6             0.520         0.15             2.1      0.054   
73            8.3             0.675         0.26             2.1      0.084   

    free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
18                  6.0                  29.0   0.9974  3.38       0.50   
38                  7.0                  19.0   0.9940  3.50       0.48   
41                 17.0                  46.0   0.9976  3.26       0.51   
45                  8.0                  65.0   0.9934  3.90       0.56   
73                 11.0                  43.0   0.9976  3.31       0.53   

    alcohol  
18      9.0  
38      9.8  
41      9.3  
45     13.1  
73      9.2

# Create a Train Test Split

Use `quality` for the y values

In [19]:
from sklearn.model_selection import train_test_split
#80% train, 20% test. y is this one column
y = normdf[['bin_quality']]
#random state 42 will have same picks for x test and y test
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

In [20]:
X_train.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
158             5.2              0.44         0.04             1.4      0.036   
304             7.3              0.41         0.24             6.8      0.057   
2952            6.1              0.27         0.25             1.8      0.041   
3414            8.9              0.30         0.35             4.6      0.032   
703             7.5              0.38         0.48             2.6      0.073   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
158                  43.0                 119.0  0.98940  3.36       0.33   
304                  41.0                 163.0  0.99490  3.20       0.41   
2952                  9.0                 109.0  0.99290  3.08       0.54   
3414                 32.0                 148.0  0.99458  3.15       0.45   
703                  22.0                  84.0  0.99720  3.32       0.70   

      alcohol  
158      12.1  
304       9.9  
2952      9.0  
3414     11.5  
703       9.6

# Train the Model using Random Forest 



In [21]:
#Train the model
from sklearn import tree
from sklearn.ensemble import RandomForestClassifier
rfc = RandomForestClassifier()
rfc.fit(X_train, y_train)

/opt/anaconda3/envs/PythonData/lib/python3.7/site-packages/ipykernel_launcher.py:5: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  """


RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=100,
                       n_jobs=None, oob_score=False, random_state=None,
                       verbose=0, warm_start=False)

In [22]:
#Test the model
from sklearn.metrics import confusion_matrix, classification_report
predictions = rfc.predict(X_test)
print(classification_report(y_test, predictions))
print(predictions[9:20])
print(y_test[9:20])

              precision    recall  f1-score   support

           0       0.62      0.65      0.63        57
           1       0.46      0.49      0.48        69
           2       0.26      0.14      0.19        69
           3       0.49      0.71      0.58        51

    accuracy                           0.48       246
   macro avg       0.46      0.50      0.47       246
weighted avg       0.45      0.48      0.45       246

[3 0 2 1 3 2 0 1 3 3 1]
     bin_quality
1610           2
1183           1
951            2
1740           2
1832           3
1385           1
147            0
4497           1
316            2
1605           3
1029           0


# Save the Model

In [20]:
# save your model by updating "your_name" with your name
# and "your_model" with your model variable
# be sure to turn this in to BCS
# if joblib fails to import, try running the command to install in terminal/git-bash
import joblib
filename = 'unbalanced-with-color.sav'
joblib.dump(predictions, filename)

['unbalanced-with-color.sav']